KAFKA-17448: New consumer seek should update positions in background thread

FrankYang0529 opened the pull request:
The `AsyncKafkaConsumer#seek` uses `SubscriptionState#seekUnvalidated` which calls `SubscriptionState#assignedState`. If we call it in app thread, it may have race condition with background thread. Move `SubscriptionState#seekUnvalidated` to background thread to avoid concurrent write.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/SeekUnvalidatedEvent.java:
```diff
@@ -0,0 +1,44 @@
+
+public class SeekUnvalidatedEvent extends CompletableApplicationEvent<Void> {
+    private final TopicPartition partition;
+    private final SubscriptionState.FetchPosition position;
```
It's just my preference, but I'd rather not pass the `FetchPosition` around in the event. Could we instead include the following directly in the event:

* `TopicPartition partition`
* `long offset`
* `Optional<Integer> offsetEpoch`

The `ApplicationEventProcessor` has access to the `ConsumerMetadata` object, so we could fetch the current leader there, in one place, instead of having it twice in the `AsyncKafkaConsumer`.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java:
```diff
@@ -409,4 +413,13 @@ protected ApplicationEventProcessor create() {
+
+    private void process(final SeekUnvalidatedEvent event) {
+        try {
+            subscriptions.seekUnvalidated(event.partition(), event.position());
```
If we update the event class, we can then move more of the logic here:

```suggestion
            SubscriptionState.FetchPosition newPosition = new SubscriptionState.FetchPosition(
                event.offset(),
                event.offsetEpoch(),
                metadata.currentLeader(event.partition())
            );
            subscriptions.seekUnvalidated(event.partition(), newPosition);
```

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -820,7 +823,10 @@ public void seek(TopicPartition partition, OffsetAndMetadata offsetAndMetadata)
-            subscriptions.seekUnvalidated(partition, newPosition);
+
+            Timer timer = time.timer(defaultApiTimeoutMs);
+            SeekUnvalidatedEvent seekUnvalidatedEventEvent = new SeekUnvalidatedEvent(calculateDeadlineMs(timer), partition, newPosition);
+            applicationEventHandler.addAndGet(seekUnvalidatedEventEvent);
```
Per the other suggestions, we could make this a bit more succinct:

```suggestion
            SeekUnvalidatedEvent event = new SeekUnvalidatedEvent(
                calculateDeadlineMs(timer),
                partition,
                offsetAndMetadata.offset(),
                offsetAndMetadata.leaderEpoch()
            );
            applicationEventHandler.addAndGet(event);
```

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -793,7 +794,9 @@ public void seek(TopicPartition partition, long offset) {
                 metadata.currentLeader(partition));
-            subscriptions.seekUnvalidated(partition, newPosition);
+            Timer timer = time.timer(defaultApiTimeoutMs);
+            SeekUnvalidatedEvent seekUnvalidatedEventEvent = new SeekUnvalidatedEvent(calculateDeadlineMs(timer), partition, newPosition);
+            applicationEventHandler.addAndGet(seekUnvalidatedEventEvent);
```
Per the other suggestions, we could defer the creation of the `SubscriptionState.FetchPosition ` until the `ApplicationEventProcessor` processor method to make this call a bit more succinct:

```suggestion
            SeekUnvalidatedEvent event = new SeekUnvalidatedEvent(
                calculateDeadlineMs(timer),
                partition,
                offset,
                Optional.empty(), // This will ensure we skip validation
            );
            applicationEventHandler.addAndGet(event);
```

kirktrue reviewed (changes requested):
Thanks for the PR, @FrankYang0529!

I think the implementation is sound. I just have the one suggestion to defer the creation of the `FetchPosition` until the `ApplicationEventProcessor.process()` invocation. This makes the calls in `AsyncKafkaConsumer` cleaner (IMO), and also keeps the call to `ConsumerMetadata.currentLeader()` as close as possible to the `seekUnvalidated()` call.

Thanks!

kirktrue commented:
Also, could you bit a sentence or two description in the event class’ Javadoc? Thanks!

FrankYang0529 commented:
@kirktrue, thanks for reviewing and suggestions. I addressed all comments and add Javadoc for the `SeekUnvalidatedEvent`. Could you take a look when you have time? Thank you.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/SeekUnvalidatedEvent.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    protected String toStringBase() {
+        return super.toStringBase() + ", partition=" + partition + ", offset=" + offset + ", offsetEpoch=" + offsetEpoch;
```
given that offsetEpoch is an Optional object, won't this print something like "offsetEpoch=Optional(10)" or "offsetEpoch=Optional.empty"? If so, I find it confusing. 

We could check offsetEpoch.isPresent, and only add "offsetsEpoch=.." to the output string if we have a value for it, seems clearer as output. What do you think?

FrankYang0529 replied:
Thanks for the review. Updated it.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:
```diff
@@ -183,6 +186,38 @@ public void testAssignmentChangeEventWithException() {
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
```
do we get any value from these params? I expect that having or not a group ID does not change the course of this test that is only about seeking to a position

FrankYang0529 replied:
Yes, I think we don't need to test both situations for this case. Thanks.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:
```diff
@@ -183,6 +186,38 @@ public void testAssignmentChangeEventWithException() {
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
```
ditto

lianetm commented:
Hey @FrankYang0529 , thanks for the patch! Took a full pass, just some minor comments.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/SeekUnvalidatedEvent.java:
```diff
@@ -0,0 +1,58 @@
+    protected String toStringBase() {
+        return offsetEpoch
+                .map(integer -> super.toStringBase() + ", partition=" + partition + ", offset=" + offset + ", offsetEpoch=" + integer)
+                .orElseGet(() -> super.toStringBase() + ", partition=" + partition + ", offset=" + offset);
```
```suggestion
        return super.toStringBase() 
            + ", partition=" + partition 
            + ", offset=" + offset 
            + offsetEpoch.map(integer -> ", offsetEpoch=" + integer).orElse("");
```
What about this, just to avoid duplicating what's common?

FrankYang0529 replied:
Agree, I think it's better solution. Thanks for the suggestion. Updated it.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/SeekUnvalidatedEvent.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Optional;
+
+/**
+ * Event to perform {@link org.apache.kafka.clients.consumer.internals.SubscriptionState#seekUnvalidated(TopicPartition, SubscriptionState.FetchPosition)}
```
one more detail, we already import SubscriptionState on ln 19, so we can simplify here to ...link SubscriptionState#....

FrankYang0529 replied:
Updated it. Thank you.

FrankYang0529 committed:
KAFKA-17448: New consumer seek should update positions in background thread

Signed-off-by: PoAn Yang <payang@apache.org>

lianetm reviewed (approved):
LGTM. Thanks!
